### PR TITLE
Fix and enhancements for OVN plugins

### PR DESCRIPTION
**Pull request by danalsan**
[ovn_host] Add support for containerized setups
[ovn_central] Add support to containerized setups
Allow reading commands output from containers
Minor enhancements to the OVN plugin

### CODE REVIEW

File: sos/plugins/ovn_central.py [hunk 1]
```diff
@@ -57,20 +76,39 @@ def setup(self):
...
+            'ovn-nbctl show',
+            'ovn-sbctl show',
             'ovn-sbctl lflow-list',
             'ovn-nbctl get-ssl',
             'ovn-nbctl get-connection',
             'ovn-sbctl get-ssl',
             'ovn-sbctl get-connection',
         ]
 
+
```
Comment (pmoravec) [inline sos/plugins/ovn_central.py]: Please remove this extra empty line (Travis fails on it - you can run "pycodestyle" on modified code to spot it locally before committing it).

File: sos/plugins/ovn_central.py [hunk 2]
```diff
@@ -57,20 +76,39 @@ def setup(self):
...
         schema_dir = '/usr/share/openvswitch'
+        if containerized:
+            for db in ['nb', 'sb']:
+                cmd = ("%s cp %s:/usr/share/openvswitch/ovn-%s.ovsschema "
+                       "/tmp/" % (self._container_runtime,
+                                  self._container_name, db))
+                if self.get_command_output(cmd)['status'] != 0:
```
Comment (pmoravec) [inline sos/plugins/ovn_central.py]: What is the purpose to copy the files? (to get the copy command output that is usefull for some troubleshooting?)

Is it safe to expose that file to publicly available `/tmp` directory - can't it be misused by some intruder?

Comment (danalsan) [reply sos/plugins/ovn_central.py]: > What is the purpose to copy the files? (to get the copy command output that is usefull for some troubleshooting?)
> 
> Is it safe to expose that file to publicly available `/tmp` directory - can't it be misused by some intruder?
> 
> (I think I got this; you collect the file in add_database_output and I guess it is really safe to expose them since non-containerised environment exposes them in similar way - am I right?)

Yes, you're right. On non containerized environments, this is exposed in the host filesystem while in containerized environments we want to read the files to figure out what tables to pull data from.

Re. security, these files are just the DB schema and are publicly available anyways in the OVS/OVN repositories, they do not contain any data but just the schema definition.

File: sos/plugins/ovn_central.py [hunk 3]
```diff
@@ -65,12 +86,37 @@ def setup(self):
...
+            for db in ['nb', 'sb']:
+                cmdline = ("%s/usr/share/openvswitch/ovn-%s.ovsschema "
+                           "/tmp/" % (cmd, db))
+                if self.get_command_output(cmdline)['status'] != 0:
+                    self._log_error("Could not retrieve DB schema file from "
+                                    "container %s", self._container_name)
+            schema_dir = '/tmp'
+            if self._container_runtime == 'podman':
+                self.get_command_output(
+                    "podman umount %s" % self._container_name)
```
Comment (TurboTurtle) [inline sos/plugins/ovn_central.py]: Absolute nack from me. In addition to the fact that we shouldn't be mounting and unmounting resources in a plugin, there is also no reason to be capturing the output of that mount/unmount operation.

File: sos/plugins/ovn_central.py [hunk 4]
```diff
@@ -65,12 +86,37 @@ def setup(self):
         ]
 
         schema_dir = '/usr/share/openvswitch'
+        if containerized:
+            if self._container_runtime == 'podman':
+                dst = self.get_command_output(
+                    "podman mount %s" % self._container_name)['output']
+                cmd = "cp " + dst.rstrip()
```
[the remark this answers is not in the document]
Comment (danalsan) [reply sos/plugins/ovn_central.py]: > `podman cp` exists and matches the `docker` syntax to copy files out of a container. In no way should we be mounting anything anywhere in a plugin.

I took a RHEL 8.0 base system where `podman cp` doesn't exist. Do you mean is it fair to assume that whatever version podman is now packaged in RHEL8 would support that command? If so, I'll fallback to my previous version.

[root@controller-0 ovn_central]# podman cp
Command "cp" not found.
See `podman --help`.


> 
> This whole block to construct the command can be reduced to:
> 
> ```python
> if containerized:
>     cmd = "%s cp %s:" % (self._container_runtime, self._container_name)
> ```

File: sos/plugins/ovn_central.py (hunk 3 above)
Comment (danalsan) [reply sos/plugins/ovn_central.py]: > Absolute nack from me. In addition to the fact that we shouldn't be mounting and unmounting resources in a plugin, there is also no reason to be capturing the output of that mount/unmount operation.

There is absolutely a reason to capture the output of the *mount* operation:

# podman mount ovn-dbs-bundle-podman-0
/var/lib/containers/storage/overlay/d1679d89311b4a4fc4909b3126ab6abf1458fd4c1968061105b788867c6da69b/merged

It gives you the location where the filesystem is (actually this is not even a mount as without executing a `podman mount` one can access the filesystem directly in the host). But, again, reading the output would be needed.

For unmouting, how would you suggest doing it if it's not through `get_command_output`?

File: sos/plugins/ovn_central.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TurboTurtle) [reply sos/plugins/ovn_central.py]: It looks like `podman cp` is a 1.1 feature, whereas 8.0 has 1.0 podman. I'd imagine the next rebase for podman in RHEL will update it to at least 1.1 (Fedora is on 1.4 right now for reference).

In any event, I am still completely against doing mount operations within a plugin. Maybe there needs to be some logic to gate against podman-1.0, or we just accept that it won't work for that version of podman.

**Comment by danalsan** [comment]
> On 22 Aug 2019, at 22:01, Jake Hunsaker <notifications@github.com> wrote:
> 
> @TurboTurtle commented on this pull request.
> 
> In sos/plugins/ovn_central.py:
> 
> > @@ -65,12 +86,37 @@ def setup(self):
>          ]
>  
>          schema_dir = '/usr/share/openvswitch'
> +        if containerized:
> +            if self._container_runtime == 'podman':
> +                dst = self.get_command_output(
> +                    "podman mount %s" % self._container_name)['output']
> +                cmd = "cp " + dst.rstrip()
> It looks like podman cp is a 1.1 feature, whereas 8.0 has 1.0 podman. I'd imagine the next rebase for podman in RHEL will update it to at least 1.1 (Fedora is on 1.4 right now for reference).
> 
> In any event, I am still completely against doing mount operations within a plugin. Maybe there needs to be some logic to gate against podman-1.0, or we just accept that it won't work for that version of podman.
> 
As I said, the directory is accesible even without mounting but I need to find it some how; it is not a real mount and anyways it’s just mounted right after the file is fetched. 

What then about using glob if Python 3 is installed to search for the files in /var/lib and “docker cp” otherwise?

We need the ovn plugin to work on RHEL8. 
Can you suggest something that would work for you and later on when podman > 1.0 is available, we will just use “podman cp”?

File: sos/plugins/ovn_central.py (hunk 3 above)
Comment (TurboTurtle) [reply sos/plugins/ovn_central.py]: Ah, fair point - I was reading that as `get_cmd_output` and not `get_command_output`, my fault there.

**Comment by TurboTurtle** [comment]
> As I said, the directory is accesible even without mounting but I need to find it some how; it is not a real mount and anyways it’s just mounted right after the file is fetched. What then about using glob if Python 3 is installed to search for the files in /var/lib and “docker cp” otherwise? We need the ovn plugin to work on RHEL8. Can you suggest something that would work for you and later on when podman > 1.0 is available, we will just use “podman cp”?


As I said, I am still not onboard with performing file copies on the host. `podman cp` is a last resort in my mind, if we even agree to go that route at all.

What is represented by the table names in the `ovn-(nb|sb).ovsschema` files? Do they map to some other component within OVN that can be listed with some other ovn-ctl command? I would find it odd if an end user needed to go digging through schema files in order to run valid `ovn-nbctl list $schema` on their own. 

If some kind of list command like that is present, we can simplify this substantially and making it container-compatible would be trivial.

**Comment by danalsan** [comment]
@TurboTurtle I changed the code to read the file inside the container to memory. If it's not a containerized setup, the old behavior is kept. This way we don't mount or copy any files into memory and it shouldn't be a problem as the schema file shouldn't grow much (it doesn't contain any data but just the tables, columns and relationships). Let me know if this works for you.
Thanks,
daniel

**Comment by bmr-cymru** [comment]
This pull request **introduces 2 alerts** when merging 8178714c550458054a3d1fa94f7be218abea1fe8 into f9e0a231495adad0e165e908dd0dde768ed62ed0 - [view on LGTM.com](https://lgtm.com/projects/g/sosreport/sos/rev/pr-4f2faa7b7d6089e72d9710375d224930ba29462c)

**new alerts:**

* 1 for Unused local variable
* 1 for Wrong number of arguments in a call

---

**Warning** - Automated code review for sosreport/sos will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by danalsan** [comment]
The travis failures are unrelated to this PR, the offending lines belong to the following commit:

^2453c238 (Jake Hunsaker 2019-06-07 13:08:12 -0400  926)                         strfile = file_name.replace(os.path.sep, ".") + ".tailed"
^2453c238 (Jake Hunsaker 2019-06-07 13:08:12 -0400  927)                         self.add_string_as_file(tail(_file, sizelimit), strfile)

I can send a separate PR to fix it.

**Comment by danalsan** [comment]
> The offending commit is actually [9c421f8](https://github.com/sosreport/sos/commit/9c421f819fcb40877aa873719656b8b3b6590d06)
> 
> But we can get that cleaned up in master in a moment before we close 3.8.

ack, thanks a lot.
Let me know please if there's anything you'd like me to change from this PR.

File: sos/utilities.py [hunk 5]
```diff
@@ -147,7 +147,7 @@ def _child_prep_fn():
         else:
             expanded_args.append(arg)
     try:
-        p = Popen(expanded_args, shell=False, stdout=PIPE,
+        p = Popen(expanded_args, stdin=PIPE, shell=False, stdout=PIPE,
```
Comment (pmoravec) [inline sos/utilities.py]: Why this change, please?

@bmr-cymru and @TurboTurtle , any comment here? Could it have some drawbacks?

File: sos/plugins/ovn_central.py [hunk 6]
```diff
@@ -19,30 +19,61 @@ class OVNCentral(Plugin):
...
+                    except Exception:
+                        self._log_error(
+                            "Cannot parse JSON file %s" % filename)
+                        return
+            except IOError as ex:
+                self._log_error(
+                    "Could not open DB schema file %s: %s" % (filename, ex))
+                return
+        return [
+            table for table in six.iterkeys(db['tables']) if table not in skip]
```
Comment (pmoravec) [inline sos/plugins/ovn_central.py]: Theoretically, the JSON file might lack 'tables' (can we 100% guarantee the json file cant be corrupted or empty?). If this would happen, an uncaught exception would be raised.

File: sos/plugins/ovn_central.py [hunk 7]
```diff
@@ -66,10 +99,20 @@ def setup(self):
...
-                                 cmds, 'ovn-nbctl')
-        self.add_database_output(os.path.join(schema_dir, 'ovn-sb.ovsschema'),
-                                 cmds, 'ovn-sbctl', ['Logical_Flow'])
+        nb_tables = self.get_tables_from_schema(os.path.join(
+            schema_dir, 'ovn-nb.ovsschema'))
+        sb_tables = self.get_tables_from_schema(os.path.join(
+            schema_dir, 'ovn-sb.ovsschema'), ['Logical_Flow'])
+
+        self.add_database_output(nb_tables, cmds, 'ovn-nbctl')
+        self.add_database_output(sb_tables, cmds, 'ovn-sbctl')
```
Comment (pmoravec) [inline sos/plugins/ovn_central.py]: Will be "cmds" variable really updated in the `cmds.append('%s list %s' % (ovn_cmd, table))` call inside `add_database_output`? As:

    >>> def plus(a,b,c):
    ...   c = a+b
    ... 
    >>> s1=3
    >>> s2=5
    >>> s=6
    >>> plus(s1,s2,s)
    >>> s
    6

(this could be over-worked by using "self.cmds" in `setup` and `add_database_output` and not passing "cmds" as its argument)

File: sos/utilities.py (hunk 5 above)
Comment (danalsan) [reply sos/utilities.py]: Thanks a lot for taking a look.
Without that change, podman hangs for me [0] and with strace I could see that the process is receiving SIGTTIN all the time.

[0] https://bugzilla.redhat.com/show_bug.cgi?id=1732525

File: sos/plugins/ovn_central.py (hunk 6 above)
Comment (danalsan) [reply sos/plugins/ovn_central.py]: Good point, plus it was handled before my patch. Will fix.

Comment (danalsan) [reply sos/plugins/ovn_central.py]: Good point, plus it was handled before my patch. Will fix.
Thanks!

File: sos/plugins/ovn_central.py (hunk 7 above)
Comment (danalsan) [reply sos/plugins/ovn_central.py]: As it is a list, we can be sure that the cmds variable will be updated. (My patch is not changing the behavior though, it was like that before). For example:

def _bb(param):
    param.append(3)

aa = [1, 2]
_bb(aa)
print(aa)

$ python3.7 test.py
[1, 2, 3]

$ python2.7 test.py
[1, 2, 3]

Comment (pmoravec) [reply sos/plugins/ovn_central.py]: Ah right, forgot it is a list..
